File: Sources/App/Classes/IRC/IRCChannelConfig.m
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2008 - 2010 Satoshi Nakagawa <psychs AT limechat DOT net>
 * Copyright (c) 2010 - 2020 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

#import "NSObjectHelperPrivate.h"
#import "TPCPreferencesLocalPrivate.h"
#import "IRCChannelConfigInternal.h"

NS_ASSUME_NONNULL_BEGIN

@implementation IRCChannelConfig

#pragma mark -
#pragma mark Defaults

- (void)populateDefaultsPreflight
{	
	if (self.initializedAsCopy) {
		return;
	}

	self->_defaults = @{
	  @"autoJoin" : @(YES),
	  @"channelType" : @(IRCChannelTypeChannel),
	  @"ignoreGeneralEventMessages"	: @(NO),
	  @"ignoreHighlights" : @(NO),
	  @"inlineMediaEnabled" : @(NO),
	  @"inlineMediaDisabled" : @(NO),
	  @"pushNotifications" : @(YES),
	  @"showTreeBadgeCount" : @(YES)
	};
}

- (void)populateDefaultsPostflight
{	
	if (self.initializedAsCopy) {
		return;
	}

	SetVariableIfNil(self->_channelName, @"")

	SetVariableIfNil(self->_uniqueIdentifier, [NSString stringWithUUID])

	SetVariableIfNil(self->_notificationsMutable, [NSMutableDictionary dictionary])
}

- (void)populateDefaultsByAppendingDictionary:(NSDictionary<NSString *, id> *)defaultsToAppend
{
	NSParameterAssert(defaultsToAppend != nil);

	self->_defaults = [self->_defaults dictionaryByAddingEntries:defaultsToAppend];
}

#pragma mark -
#pragma mark Channel Configuration

+ (IRCChannelConfig *)seedWithName:(NSString *)channelName
{
	NSParameterAssert(channelName != nil);

	NSDictionary *dic = @{@"channelName" : channelName};

	IRCChannelConfig *config = [[self alloc] initWithDictionary:dic];

	return config;
}

- (instancetype)init
{
	return [super initWithDictionary:@{}];
}

- (void)initializedClassHealthCheck
{
	if (self.mutable || self.initializedAsCopy) {
		return;
	}

	NSParameterAssert(self->_channelName.length > 0);
}

- (void)populateDictionaryValues:(NSDictionary<NSString *, id> *)dic
{
	NSParameterAssert(dic != nil);

	NSMutableDictionary<NSString *, id> *defaultsMutable = [self->_defaults mutableCopy];

	[defaultsMutable addEntriesFromDictionary:dic];

	[defaultsMutable assignBoolTo:&self->_pushNotifications forKey:@"pushNotifications"];
	[defaultsMutable assignBoolTo:&self->_showTreeBadgeCount forKey:@"showTreeBadgeCount"];

	[defaultsMutable assignStringTo:&self->_channelName forKey:@"channelName"];
	[defaultsMutable assignStringTo:&self->_uniqueIdentifier forKey:@"uniqueIdentifier"];

	[defaultsMutable assignUnsignedIntegerTo:&self->_type forKey:@"channelType"];

	if (self->_type != IRCChannelTypeChannel) {
		return;
	}

	/* Load the newest set of keys */
	[defaultsMutable assignBoolTo:&self->_autoJoin forKey:@"autoJoin"];
	[defaultsMutable assignBoolTo:&self->_ignoreGeneralEventMessages forKey:@"ignoreGeneralEventMessages"];
	[defaultsMutable assignBoolTo:&self->_ignoreHighlights forKey:@"ignoreHighlights"];
	[defaultsMutable assignBoolTo:&self->_inlineMediaDisabled forKey:@"inlineMediaDisabled"];
	[defaultsMutable assignBoolTo:&self->_inlineMediaEnabled forKey:@"inlineMediaEnabled"];

	[defaultsMutable assignStringTo:&self->_label forKey:@"label"];

	[defaultsMutable assignStringTo:&self->_defaultModes forKey:@"defaultMode"];
	[defaultsMutable assignStringTo:&self->_defaultTopic forKey:@"defaultTopic"];

	NSDictionary *notifications = [defaultsMutable dictionaryForKey:@"notifications"];

	if (notifications != nil) {
		self->_notificationsMutable = [notifications mutableCopy];
	}

	/* Load legacy keys (if they exist) */
	if (self.initializedAsCopy) {
		return;
	}

	[defaultsMutable assignBoolTo:&self->_autoJoin forKey:@"joinOnConnect"];
	[defaultsMutable assignBoolTo:&self->_ignoreGeneralEventMessages forKey:@"ignoreJPQActivity"];
	[defaultsMutable assignBoolTo:&self->_pushNotifications forKey:@"enableNotifications"];
	[defaultsMutable assignBoolTo:&self->_showTreeBadgeCount forKey:@"enableTreeBadgeCountDrawing"];

	/* Migrate inline media */
	/* Old behavior was to store a single property named "ignoreInlineMedia"
	 Depending on the value of the global preference, the value of this
	 property was used to determine whether to hide inline media per-channel
	 or to show it per-channel. That is stupid idea because if someone
	 has it enabled globally, has it turned off in a channel, turns it
	 off globally, then it is turned on in that channel. We split it
	 up into two properties and this logic performs migration. */
	{
		/* Do new keys exist in incoming dictionary/ */
		if (dic[@"inlineMediaEnabled"] != nil &&
			dic[@"inlineMediaDisabled"] != nil)
		{
			return;
		}

		NSNumber *ignoreInlineMedia = dic[@"ignoreInlineMedia"]; // old key

		/* If old value is NO, then we do not have to continue
		 because the default value for the new values is NO. */
		if (ignoreInlineMedia == nil || ignoreInlineMedia.boolValue == NO) {
			return;
		}

		BOOL inlineEnabledGlobally = [TPCPreferences showInlineMedia];

		/* Old property was the inverse of the global */
		/* Global enabled = local disabled,
		   Global disabled = local enabled */
		self->_inlineMediaDisabled = inlineEnabledGlobally;
		self->_inlineMediaEnabled = !inlineEnabledGlobally;
	}
}

- (NSDictionary<NSString *, id> *)dictionaryValueForTarget:(XRPortablePropertyDictTarget)target
{
	NSMutableDictionary<NSString *, id> *dic = [NSMutableDictionary dictionary];

	[dic setBool:self.pushNotifications	forKey:@"pushNotifications"];
	[dic setBool:self.showTreeBadgeCount forKey:@"showTreeBadgeCount"];

	if (self.type == IRCChannelTypeChannel) {
		[dic maybeSetObject:self.label forKey:@"label"];

		[dic maybeSetObject:self.defaultModes forKey:@"defaultMode"];
		[dic maybeSetObject:self.defaultTopic forKey:@"defaultTopic"];

		[dic maybeSetObject:self.notifications forKey:@"notifications"];

		[dic setBool:self.autoJoin forKey:@"autoJoin"];

		[dic setBool:self.ignoreGeneralEventMessages forKey:@"ignoreGeneralEventMessages"];
		[dic setBool:self.ignoreHighlights forKey:@"ignoreHighlights"];

		[dic setBool:self.inlineMediaDisabled forKey:@"inlineMediaDisabled"];
		[dic setBool:self.inlineMediaEnabled forKey:@"inlineMediaEnabled"];
	}

	[dic maybeSetObject:self.channelName forKey:@"channelName"];
	[dic maybeSetObject:self.uniqueIdentifier forKey:@"uniqueIdentifier"];

	[dic setUnsignedInteger:self.type forKey:@"channelType"];

	if (target == XRPortablePropertyDictTargetCopy ||
		target == XRPortablePropertyDictTargetMutableCopy)
	{
		return [dic copy];
	}

	return [dic dictionaryByRemovingDefaults:self->_defaults allowEmptyValues:YES];
}

- (BOOL)isEqual:(id)object
{
	if (object == nil) {
		return NO;
	}

	if (object == self) {
		return YES;
	}

	if ([object isKindOfClass:[IRCChannelConfig class]] == NO) {
		return NO;
	}

	IRCChannelConfig *objectCast = (IRCChannelConfig *)object;

	NSDictionary *s1 = self.dictionaryValue;

	NSDictionary *s2 = objectCast.dictionaryValue;

	return ([s1 isEqualToDictionary:s2] &&
			((self->_secretKey == nil && objectCast->_secretKey == nil) ||
			 [self->_secretKey isEqualToString:objectCast->_secretKey]));
}

- (NSUInteger)hash
{
	return self.uniqueIdentifier.hash;
}

- (id)copyAsMutable:(BOOL)mutableCopy uniquing:(BOOL)uniquing
{
	IRCChannelConfig *config = [self allocForCopyAsMutable:mutableCopy];

	config->_defaults = self->_defaults;

	config->_secretKey = self->_secretKey;

	if (uniquing) {
		config->_uniqueIdentifier = [NSString stringWithUUID];
	}

	return [config initWithDictionary:self.dictionaryValueForCopy];
}

- (__kindof XRPortablePropertyDict *)mutableClass
{
	return [IRCChannelConfigMutable self];
}

- (NSDictionary<NSString *, NSNumber *> *)notifications
{
	@synchronized (self->_notificationsMutable) {
		return [self->_notificationsMutable copy];
	}
}

#pragma mark -
#pragma mark Keychain Management

- (nullable NSString *)secretKey
{
	if (self->_secretKey) {
		return self->_secretKey;
	}

	return self.secretKeyFromKeychain;
}

- (nullable NSString *)secretKeyFromKeychain
{
	NSString *secretKeyServiceName = [NSString stringWithFormat:@"glasstual.cjoinkey.%@", self.uniqueIdentifier];

	NSString *kcPassword = [XRKeychain getPasswordFromKeychainItem:@"Glasstual (Channel JOIN Key)"
													  withItemKind:@"application password"
													   forUsername:nil
													   serviceName:secretKeyServiceName];

	return kcPassword;
}

- (void)writeSecretKeyToKeychain
{
	if (self->_secretKey == nil) {
		return;
	}

	NSString *secretKeyServiceName = [NSString stringWithFormat:@"glasstual.cjoinkey.%@", self.uniqueIdentifier];

	[XRKeychain modifyOrAddKeychainItem:@"Glasstual (Channel JOIN Key)"
						   withItemKind:@"application password"
							forUsername:nil
						withNewPassword:self->_secretKey
							serviceName:secretKeyServiceName];

	self->_secretKey = nil;
}

- (void)destroySecretKeyKeychainItem
{
	NSString *secretKeyServiceName = [NSString stringWithFormat:@"glasstual.cjoinkey.%@", self.uniqueIdentifier];

	[XRKeychain deleteKeychainItem:@"Glasstual (Channel JOIN Key)"
					  withItemKind:@"application password"
					   forUsername:nil
					   serviceName:secretKeyServiceName];

	/* Reset temporary value */
	self->_secretKey = nil;
}

#pragma mark -
#pragma mark Notifications

- (nullable NSString *)soundForEvent:(TXNotificationType)event
{
	NSString *eventKey = [TPCPreferences keyForEvent:event category:@"Sound"];

	if (eventKey == nil) {
		return nil;
	}

	/* @synchronized is used here because IRCChannelConfigMutable can modify
	 this value underneath us. */
	@synchronized (self->_notificationsMutable) {
		return self->_notificationsMutable[eventKey];
	}
}

- (NSControlStateValue)_stateForEventKey:(NSString *)eventKey
{
	NSParameterAssert(eventKey != nil);

	@synchronized (self->_notificationsMutable) {
		NSNumber *value = self->_notificationsMutable[eventKey];

		if (value == nil) {
			return NSControlStateValueMixed;
		}

		if (value.boolValue == NO) {
			return NSControlStateValueOff;
		}

		return NSControlStateValueOn;
	}
}

- (NSControlStateValue)notificationEnabledForEvent:(TXNotificationType)event
{
	NSString *eventKey = [TPCPreferences keyForEvent:event category:@"Enabled"];

	if (eventKey == nil) {
		return NO;
	}

	return [self _stateForEventKey:eventKey];
}

- (NSControlStateValue)disabledWhileAwayForEvent:(TXNotificationType)event
{
	NSString *eventKey = [TPCPreferences keyForEvent:event category:@"Disable While Away"];

	if (eventKey == nil) {
		return NO;
	}

	return [self _stateForEventKey:eventKey];
}

- (NSControlStateValue)bounceDockIconForEvent:(TXNotificationType)event
{
	NSString *eventKey = [TPCPreferences keyForEvent:event category:@"Bounce Dock Icon"];

	if (eventKey == nil) {
		return NO;
	}

	return [self _stateForEventKey:eventKey];
}

- (NSControlStateValue)bounceDockIconRepeatedlyForEvent:(TXNotificationType)event
{
	NSString *eventKey = [TPCPreferences keyForEvent:event category:@"Bounce Dock Icon Repeatedly"];

	if (eventKey == nil) {
		return NO;
	}

	return [self _stateForEventKey:eventKey];
}

- (NSControlStateValue)speakEvent:(TXNotificationType)event
{
	NSString *eventKey = [TPCPreferences keyForEvent:event category:@"Speak"];

	if (eventKey == nil) {
		return NO;
	}

	return [self _stateForEventKey:eventKey];
}

@end

#pragma mark -

@implementation IRCChannelConfigMutable

@dynamic type;
@dynamic autoJoin;
@dynamic channelName;
@dynamic label;
@dynamic defaultModes;
@dynamic defaultTopic;
@dynamic ignoreGeneralEventMessages;
@dynamic ignoreHighlights;
@dynamic inlineMediaDisabled;
@dynamic inlineMediaEnabled;
@dynamic pushNotifications;
@dynamic secretKey;
@dynamic showTreeBadgeCount;

+ (BOOL)isMutable
{
	return YES;
}

- (__kindof XRPortablePropertyDict *)immutableClass
{
	return [IRCChannelConfig self];
}

- (void)setType:(IRCChannelType)type
{
	if (self->_type != type) {
		self->_type = type;
	}
}

- (void)setAutoJoin:(BOOL)autoJoin
{
	if (self->_autoJoin != autoJoin) {
		self->_autoJoin = autoJoin;
	}
}

- (void)setIgnoreGeneralEventMessages:(BOOL)ignoreGeneralEventMessages
{
	if (self->_ignoreGeneralEventMessages != ignoreGeneralEventMessages) {
		self->_ignoreGeneralEventMessages = ignoreGeneralEventMessages;
	}
}

- (void)setIgnoreHighlights:(BOOL)ignoreHighlights
{
	if (self->_ignoreHighlights != ignoreHighlights) {
		self->_ignoreHighlights = ignoreHighlights;
	}
}

- (void)setInlineMediaDisabled:(BOOL)inlineMediaDisabled
{
	if (self->_inlineMediaDisabled != inlineMediaDisabled) {
		self->_inlineMediaDisabled = inlineMediaDisabled;
	}
}

- (void)setInlineMediaEnabled:(BOOL)inlineMediaEnabled
{
	if (self->_inlineMediaEnabled != inlineMediaEnabled) {
		self->_inlineMediaEnabled = inlineMediaEnabled;
	}
}

- (void)setPushNotifications:(BOOL)pushNotifications
{
	if (self->_pushNotifications != pushNotifications) {
		self->_pushNotifications = pushNotifications;
	}
}

- (void)setShowTreeBadgeCount:(BOOL)showTreeBadgeCount
{
	if (self->_showTreeBadgeCount != showTreeBadgeCount) {
		self->_showTreeBadgeCount = showTreeBadgeCount;
	}
}

- (void)setChannelName:(NSString *)channelName
{
	NSParameterAssert(channelName != nil);

	if (self->_channelName != channelName) {
		self->_channelName = [channelName copy];
	}
}

- (void)setLabel:(nullable NSString *)label
{
	if (self->_label != label) {
		self->_label = [label copy];
	}
}

- (void)setDefaultModes:(nullable NSString *)defaultModes
{
	if (self->_defaultModes != defaultModes) {
		self->_defaultModes = [defaultModes copy];
	}
}

- (void)setDefaultTopic:(nullable NSString *)defaultTopic
{
	if (self->_defaultTopic != defaultTopic) {
		self->_defaultTopic = [defaultTopic copy];
	}
}

- (void)setSecretKey:(nullable NSString *)secretKey
{
	if (self->_secretKey != secretKey) {
		self->_secretKey = [secretKey copy];
	}
}

- (void)setSound:(nullable NSString *)value forEvent:(TXNotificationType)event
{
	NSString *eventKey = [TPCPreferences keyForEvent:event category:@"Sound"];

	if (eventKey == nil) {
		return;
	}

	@synchronized (self->_notificationsMutable) {
		if (value == nil) {
			[self->_notificationsMutable removeObjectForKey:eventKey];
		} else {
			self->_notificationsMutable[eventKey] = value;
		}
	}
}

- (void)_setState:(NSControlStateValue)state forEventKey:(NSString *)eventKey
{
	NSParameterAssert(eventKey != nil);

	@synchronized (self->_notificationsMutable) {
		switch (state) {
			case NSControlStateValueOn:
			{
				self->_notificationsMutable[eventKey] = @(YES);

				break;
			}
			case NSControlStateValueOff:
			{
				self->_notificationsMutable[eventKey] = @(NO);

				break;
			}
			case NSControlStateValueMixed:
			{
				[self->_notificationsMutable removeObjectForKey:eventKey];

				break;
			}
			default:
			{
				NSAssert(NO, @"Bad 'state'");
			}
		}
	}
}

- (void)setNotificationEnabled:(NSControlStateValue)value forEvent:(TXNotificationType)event
{
	NSString *eventKey = [TPCPreferences keyForEvent:event category:@"Enabled"];

	if (eventKey == nil) {
		return;
	}

	[self _setState:value forEventKey:eventKey];
}

- (void)setDisabledWhileAway:(NSControlStateValue)value forEvent:(TXNotificationType)event
{
	NSString *eventKey = [TPCPreferences keyForEvent:event category:@"Disable While Away"];

	if (eventKey == nil) {
		return;
	}

	[self _setState:value forEventKey:eventKey];
}

- (void)setBounceDockIcon:(NSControlStateValue)value forEvent:(TXNotificationType)event
{
	NSString *eventKey = [TPCPreferences keyForEvent:event category:@"Bounce Dock Icon"];

	if (eventKey == nil) {
		return;
	}

	[self _setState:value forEventKey:eventKey];
}

- (void)setBounceDockIconRepeatedly:(NSControlStateValue)value forEvent:(TXNotificationType)event
{
	NSString *eventKey = [TPCPreferences keyForEvent:event category:@"Bounce Dock Icon Repeatedly"];

	if (eventKey == nil) {
		return;
	}

	[self _setState:value forEventKey:eventKey];
}

- (void)setEventIsSpoken:(NSControlStateValue)value forEvent:(TXNotificationType)event
{
	NSString *eventKey = [TPCPreferences keyForEvent:event category:@"Speak"];

	if (eventKey == nil) {
		return;
	}

	[self _setState:value forEventKey:eventKey];
}

@end

NS_ASSUME_NONNULL_END

File: Sources/App/Protocol/Presence/IRCChannelConfig.swift
/* *********************************************************************
 *                  _____         _               _
 *                 |_   _|____  _| |_ _   _  __ _| |
 *                   | |/ _ \ \/ / __| | | |/ _` | |
 *                   | |  __/>  <| |_| |_| | (_| | |
 *                   |_|\___/_/\_\__|\__,_|\__,_|_|
 *
 * Copyright (c) 2008 - 2010 Satoshi Nakagawa <psychs AT limechat DOT net>
 * Copyright (c) 2010 - 2026 Codeux Software, LLC & respective contributors.
 *       Please see Acknowledgements.pdf for additional information.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *  * Neither the name of Textual, "Codeux Software, LLC", nor the
 *    names of its contributors may be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 *********************************************************************** */

import CocoaExtensions
import Foundation
import GlasstualPluginKit

/** A channel's answer to one notification setting.

 A channel either turns the setting on, turns it off, or says nothing and lets
 the application-wide preference answer. Keeping that choice in the protocol
 model means neither persistence nor the editor depends on AppKit controls. */
public nonisolated enum ChannelEventOverride: CaseIterable, Sendable, Equatable, Hashable { // nonisolated: value
	case on
	case off
	/// No override: whatever the application-wide preference says.
	case inherited
}

/// One notification override a channel carries: either a sound name or an
/// on/off flag. A channel with no override for an event inherits the
/// application-wide setting, which the sheets show as a mixed checkbox.
public nonisolated enum ChannelNotificationSetting: Codable, Sendable, Equatable, Hashable { // nonisolated: value
	case sound(String)
	case flag(Bool)

	public init(from decoder: any Decoder) throws {
		let container = try decoder.singleValueContainer()

		if let flag = try? container.decode(Bool.self) {
			self = .flag(flag)
		} else {
			self = try .sound(container.decode(String.self))
		}
	}

	public func encode(to encoder: any Encoder) throws {
		var container = encoder.singleValueContainer()

		switch self {
		case let .sound(name): try container.encode(name)
		case let .flag(value): try container.encode(value)
		}
	}
}

/** A channel or query as its connection stores it.

 A query keeps only its name, identifier and type; the channel settings below
 are read and written only for `.channel`, which is what earlier releases did
 and what keeps a stored configuration re-encoding unchanged. */
public nonisolated struct ChannelConfig: Codable, Sendable, Equatable, Hashable { // nonisolated: value
	public var uniqueIdentifier: String
	public var channelName: String
	public var type: ChannelType

	public var autoJoin = true
	public var ignoreGeneralEventMessages = false
	public var ignoreHighlights = false
	public var inlineMediaDisabled = false
	public var inlineMediaEnabled = false
	public var pushNotifications = true
	public var showTreeBadgeCount = true

	public var label: String?
	public var defaultModes: String?
	public var defaultTopic: String?

	/// Per-event overrides, keyed by the notification preference schema.
	public var notifications: [String: ChannelNotificationSetting] = [:]

	/** A channel key waiting to be written to the keychain, or one read back
	 out of it so a duplicate can carry it to its own identifier. Never
	 encoded — see `secretKey`. */
	public var pendingSecretKey: String?

	public init(
		uniqueIdentifier: String = UUID().uuidString,
		channelName: String = "",
		type: ChannelType = .channel
	) {
		self.uniqueIdentifier = uniqueIdentifier
		self.channelName = channelName
		self.type = type
	}

	public static func seed(withName channelName: String) -> ChannelConfig {
		ChannelConfig(channelName: channelName)
	}

	private enum CodingKeys: String, CodingKey {
		case uniqueIdentifier
		case channelName
		case channelType
		case autoJoin
		case ignoreGeneralEventMessages
		case ignoreHighlights
		case inlineMediaDisabled
		case inlineMediaEnabled
		case pushNotifications
		case showTreeBadgeCount
		case label
		case defaultModes = "defaultMode"
		case defaultTopic
		case notifications

		// Spellings written by releases before these settings were renamed.
		case joinOnConnect
		case ignoreJPQActivity
		case enableNotifications
		case enableTreeBadgeCountDrawing
		case ignoreInlineMedia
	}

	public init(from decoder: any Decoder) throws {
		let container = try decoder.container(keyedBy: CodingKeys.self)

		let identifier = container.decode(String.self, forKey: .uniqueIdentifier, aliases: [], default: "")
		uniqueIdentifier = identifier.isEmpty ? UUID().uuidString : identifier
		channelName = container.decode(String.self, forKey: .channelName, aliases: [], default: "")
		type = ChannelType(rawValue: container.decode(UInt.self, forKey: .channelType, aliases: [], default: 0))
			?? .channel
		pushNotifications = container.decode(
			Bool.self,
			forKey: .pushNotifications,
			aliases: [.enableNotifications],
			default: true
		)
		showTreeBadgeCount = container.decode(
			Bool.self,
			forKey: .showTreeBadgeCount,
			aliases: [.enableTreeBadgeCountDrawing],
			default: true
		)

		guard type == .channel else {
			return
		}

		decodeChannelSettings(from: container)
	}

	private mutating func decodeChannelSettings(from container: KeyedDecodingContainer<CodingKeys>) {
		autoJoin = container.decode(Bool.self, forKey: .autoJoin, aliases: [.joinOnConnect], default: true)
		ignoreGeneralEventMessages = container.decode(
			Bool.self,
			forKey: .ignoreGeneralEventMessages,
			aliases: [.ignoreJPQActivity],
			default: false
		)
		ignoreHighlights = container.decode(Bool.self, forKey: .ignoreHighlights, aliases: [], default: false)
		label = container.decodeOptional(String.self, forKey: .label)
		defaultModes = container.decodeOptional(String.self, forKey: .defaultModes)
		defaultTopic = container.decodeOptional(String.self, forKey: .defaultTopic)
		notifications = container.decodeOptional(
			[String: ChannelNotificationSetting].self,
			forKey: .notifications
		) ?? [:]

		decodeInlineMediaSettings(from: container)
	}

	/** Before these two settings existed a channel only recorded that it opted
	 out of inline media, which meant the opposite of the application-wide
	 setting. Both keys being present means the migration already ran. */
	private mutating func decodeInlineMediaSettings(from container: KeyedDecodingContainer<CodingKeys>) {
		let storedDisabled = container.decodeOptional(Bool.self, forKey: .inlineMediaDisabled)
		let storedEnabled = container.decodeOptional(Bool.self, forKey: .inlineMediaEnabled)

		inlineMediaDisabled = storedDisabled ?? false
		inlineMediaEnabled = storedEnabled ?? false

		guard storedDisabled == nil || storedEnabled == nil,
		      container.decodeOptional(Bool.self, forKey: .ignoreInlineMedia) == true
		else {
			return
		}

		inlineMediaDisabled = Preferences.Messages.showInlineMedia.detachedValue
		inlineMediaEnabled = inlineMediaDisabled == false
	}

	/** Writes the canonical keys only, and only those that differ from the
	 default — the same dictionary `ce_dictionaryByRemovingDefaults` produced. */
	public func encode(to encoder: any Encoder) throws {
		var container = encoder.container(keyedBy: CodingKeys.self)

		if pushNotifications == false {
			try container.encode(false, forKey: .pushNotifications)
		}

		if showTreeBadgeCount == false {
			try container.encode(false, forKey: .showTreeBadgeCount)
		}

		if type == .channel {
			try encodeChannelSettings(into: &container)
		}

		try container.encode(channelName, forKey: .channelName)
		try container.encode(uniqueIdentifier, forKey: .uniqueIdentifier)

		if type != .channel {
			try container.encode(type.rawValue, forKey: .channelType)
		}
	}

	private func encodeChannelSettings(into container: inout KeyedEncodingContainer<CodingKeys>) throws {
		try container.encodeIfPresent(label, forKey: .label)
		try container.encodeIfPresent(defaultModes, forKey: .defaultModes)
		try container.encodeIfPresent(defaultTopic, forKey: .defaultTopic)
		// Written even when empty, as it always has been.
		try container.encode(notifications, forKey: .notifications)

		if autoJoin == false {
			try container.encode(false, forKey: .autoJoin)
		}

		let flags: [(Bool, CodingKeys)] = [
			(ignoreGeneralEventMessages, .ignoreGeneralEventMessages),
			(ignoreHighlights, .ignoreHighlights),
			(inlineMediaDisabled, .inlineMediaDisabled),
			(inlineMediaEnabled, .inlineMediaEnabled),
		]

		for (value, key) in flags where value {
			try container.encode(true, forKey: key)
		}
	}
}

public nonisolated extension ChannelConfig { // nonisolated: value
	var keychainItem: KeychainItem {
		.channelSecretKey(uniqueIdentifier)
	}

	var secretKeyFromKeychain: String? {
		keychainItem.password
	}

	/// The key to JOIN with: an unflushed edit if there is one, and otherwise
	/// whatever the keychain holds.
	var secretKey: String? {
		get { pendingSecretKey ?? secretKeyFromKeychain }
		set { pendingSecretKey = newValue }
	}

	mutating func writeSecretKeyToKeychain() {
		guard let pendingSecretKey else {
			return
		}

		keychainItem.write(pendingSecretKey)
		self.pendingSecretKey = nil
	}

	mutating func destroySecretKeyKeychainItem() {
		keychainItem.delete()
		pendingSecretKey = nil
	}

	/// A copy under a fresh identity, carrying the channel key across so the
	/// duplicate does not silently lose it.
	func uniqueCopy() -> ChannelConfig {
		var copy = self
		copy.pendingSecretKey = pendingSecretKey ?? secretKeyFromKeychain
		copy.uniqueIdentifier = UUID().uuidString

		return copy
	}

	func sound(forEvent event: TXNotificationType) -> String? {
		guard case let .sound(name) = notifications[event.preferenceKeyName(for: .sound)]
		else {
			return nil
		}

		return name
	}

	func notificationEnabled(forEvent event: TXNotificationType) -> ChannelEventOverride {
		state(for: event, setting: .enabled)
	}

	func disabledWhileAway(forEvent event: TXNotificationType) -> ChannelEventOverride {
		state(for: event, setting: .disabledWhileAway)
	}

	func bounceDockIcon(forEvent event: TXNotificationType) -> ChannelEventOverride {
		state(for: event, setting: .bounceDockIcon)
	}

	func bounceDockIconRepeatedly(forEvent event: TXNotificationType) -> ChannelEventOverride {
		state(for: event, setting: .bounceDockIconRepeatedly)
	}

	func speakEvent(_ event: TXNotificationType) -> ChannelEventOverride {
		state(for: event, setting: .speak)
	}

	mutating func setSound(_ value: String?, forEvent event: TXNotificationType) {
		notifications[event.preferenceKeyName(for: .sound)] = value.map(ChannelNotificationSetting.sound)
	}

	mutating func setNotificationEnabled(_ value: ChannelEventOverride, forEvent event: TXNotificationType) {
		setState(value, forEvent: event, setting: .enabled)
	}

	mutating func setDisabledWhileAway(_ value: ChannelEventOverride, forEvent event: TXNotificationType) {
		setState(value, forEvent: event, setting: .disabledWhileAway)
	}

	mutating func setBounceDockIcon(_ value: ChannelEventOverride, forEvent event: TXNotificationType) {
		setState(value, forEvent: event, setting: .bounceDockIcon)
	}

	mutating func setBounceDockIconRepeatedly(_ value: ChannelEventOverride, forEvent event: TXNotificationType) {
		setState(value, forEvent: event, setting: .bounceDockIconRepeatedly)
	}

	mutating func setEventIsSpoken(_ value: ChannelEventOverride, forEvent event: TXNotificationType) {
		setState(value, forEvent: event, setting: .speak)
	}

	private mutating func setState(
		_ value: ChannelEventOverride,
		forEvent event: TXNotificationType,
		setting: NotificationSetting
	) {
		let key = event.preferenceKeyName(for: setting)

		switch value {
		case .on:
			notifications[key] = .flag(true)
		case .off:
			notifications[key] = .flag(false)
		case .inherited:
			notifications.removeValue(forKey: key)
		}
	}

	private func state(for event: TXNotificationType, setting: NotificationSetting) -> ChannelEventOverride {
		guard case let .flag(value) = notifications[event.preferenceKeyName(for: setting)]
		else {
			return .inherited
		}

		return value ? .on : .off
	}
}
